### PR TITLE
Fix crash if ICU default locale has BCP47 extensions. Fix ures_openDirect crash with NULL locale.

### PR DESCRIPTION
**Pull request by jefgen**
## Summary
This PR fixes two issues:
- Fix crash if ICU's default locale has BCP47 Unicode Extensions.
- Fix `ures_openDirect` crash with a NULL input locale ID.

It also adds a test case for `ures_openDirect` with a NULL input locale ID.

This change adds a new ICU-PATCH file for the changes, though hopefully this won't be needed and we can fix this issue upstream before ingesting a new version of ICU.

## PR Checklist
* [x] I have verified that my change is specific to this fork and cannot be made upstream.
  *  **Note:** I've filed upstream tickets for these two issues and will be making a PR for them later.
* [ ] I am making a maintenance related change.
* [ ] I am making a change that is related to usage internal to Microsoft.
* [ ] I am making a change that is related to the Windows OS build of ICU.
* [x] CLA signed. If not, please see [here](https://cla.opensource.microsoft.com/microsoft/icu) to sign the CLA.

#### ICU-21705 Calling ures_openDirect with NULL locale ID results in crash (undefined behavior)

Calling the public API `ures_openDirect` will crash if you call it with a `NULL` Locale ID -- even though this is permitted by the API docs.

The issue is that `strcmp` is called with the input locale ID (ie: null pointer), which is undefined behavior, and causes the crash.

This doesn't occur with `ures_open`, as that calls `uloc_getBaseName()` first, which gets the default locale via `_canonicalize()`, so that the call to `strcmp` has a locale ID (rather than a null pointer).

We need to check the input locale ID and adjust it accordingly if it is null (default locale) or a pointer to empty string (root locale). 

#### ICU-21706 ICU4C test suite crashes if the default locale has any BCP47 Unicode extension tags on it (ex: "en-US-u-hc-12")

If ICU's default locale has any BCP47 Unicode extension tags on it (ex: "en-US-u-hc-12") then the ICU test suite will crash.

The problem is that the `resbMutex` is attempted to be locked twice, which leads to the crash/termination.

The issue occurs on the first call to `ures_open()`. This causes the ICU data file to be loaded. However, the default locale isn't yet cached in `gDefaultLocale`, so `Locale::getDefault` needs to query the host OS for the default locale.

If the default locale has any BCP47 Unicode extension tags, this causes `_canonicalize` to attempt to convert them to the legacy ICU style extensions by calling `uloc_forLanguageTag`. As part of this conversion, `uloc_toLegacyKey` will also attempt to load the ICU data file, in order to load the `keyTypeData` to map the extensions.

However, this is problematic, as it means that we're now trying to load the data file while trying to load the data file. In other words, it means that the `resbMutex` will attempted to be locked twice – leading to the crash.

However, we can avoid this by moving the query for the default locale to be outside of the mutex protected part of the code – which allows us to avoid the circular nature of the issue. We can query and store the ICU default locale, and then pass it to the `findFirstExisting` function inside the mutex protected part of the code, so it doesn’t need to query for the default locale itself.

### CODE REVIEW

**Review by erik0686 — APPROVED** [review]
Thank you Jeff!

File: icu/icu4c/source/common/uresbund.cpp [hunk 1]
```diff
@@ -452,11 +452,10 @@ getPoolEntry(const char *path, UErrorCode *status) {
 /* INTERNAL: */
 /*   CAUTION:  resbMutex must be locked when calling this function! */
 static UResourceDataEntry *
-findFirstExisting(const char* path, char* name,
+findFirstExisting(const char* path, char* name, const char* defaultLocale,
```
Comment (huichen123) [inline icu/icu4c/source/common/uresbund.cpp:455]: >defaultLocale [](http://example.com/codeflow?start=60&length=13)

this is no longer defaultLocale.

Comment (jefgen) [reply icu/icu4c/source/common/uresbund.cpp:455]: not sure I understand, what else should this be called perhaps?

(this parameter is the name of the default locale)

Comment (huichen123) [reply icu/icu4c/source/common/uresbund.cpp:455]: to this function, it's just a locale you pass in. So it should just be named "locale". Here caller happens to pass in the default locale, but that's irrelevant to the function.

Comment (jefgen) [reply icu/icu4c/source/common/uresbund.cpp:455]: Hm, I'm not sure I fully agree that the function doesn't care what locale is passed in though... It uses this input locale name to set the `isDefault` out parameter (ie: it is checking if the entry found is also the default locale).
So I think the function does want this to be the name of the default locale.

Comment (jefgen) [reply icu/icu4c/source/common/uresbund.cpp:455]: you're right that we could pass in any name for any locale though, but I thought that it would help the reader if it was named defaultLocale

File: icu-patches/patches/019-ICU-Patch-ICU-21705_Fix_ures_openDirect_crash_with_NULL_localeID.patch [hunk 2]
```diff
@@ -0,0 +1,113 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
```
Comment (daniel-ju) [inline icu-patches/patches/019-ICU-Patch-ICU-21705_Fix_ures_openDirect_crash_with_NULL_localeID.patch:1]: Thanks for adding this too!

File: icu/icu4c/source/common/uresbund.cpp (hunk 1 above)
Comment (huichen123) [reply icu/icu4c/source/common/uresbund.cpp:455]: i c. tricky :). Thanks!

**Review by daniel-ju — APPROVED** [review]

File: icu/icu4c/source/common/uresbund.cpp (hunk 1 above)
Comment (jefgen) [reply icu/icu4c/source/common/uresbund.cpp:455]: yeah... the code isn't very easy to follow... :/ 
thank you very kindly for taking a look at the PR though! :)

File: icu-patches/patches/019-ICU-Patch-ICU-21705_Fix_ures_openDirect_crash_with_NULL_localeID.patch (hunk 2 above)
Comment (jefgen) [reply icu-patches/patches/019-ICU-Patch-ICU-21705_Fix_ures_openDirect_crash_with_NULL_localeID.patch:1]: Thanks! 
I thought it might be better to add this just in case we don't get the changes into v70 and we end up updating to that version... 🙃

**Review by axelandrejs — APPROVED** [review]
